src/segment_reshape/map_tool/segment_reshape_tool.py

```diff
@@ -158,45 +170,64 @@ def _handle_mouse_click_event(
                 success=True,
             )
 
+    def _handle_key_event(self, key: Qt.Key) -> None:
+        if self.reshape_mode is True:
+            if key == Qt.Key_Escape:
+                self.new_segment_rubber_band.reset()
+                self.temporary_new_segment_rubber_band.reset()
+                self._change_to_pick_location_mode()
+                return
+            elif key in (Qt.Key_Backspace, Qt.Key_Delete):
+                self.temporary_new_segment_rubber_band.reset()
+                if self.new_segment_rubber_band.numberOfVertices() > 1:
+                    self.new_segment_rubber_band.removeLastPoint(0, True)
+                    self.temporary_new_segment_rubber_band.addPoint(
+                        self.new_segment_rubber_band.getPoint(
+                            0, self.new_segment_rubber_band.numberOfVertices() - 1
+                        )
+                    )
+                    self.temporary_new_segment_rubber_band.addPoint(self.cursor_point)
+                else:
+                    self.new_segment_rubber_band.reset()
+                    self.temporary_new_segment_rubber_band.addPoint(
+                        QgsPointXY(self.start_point), True
+                    )
+                    self.temporary_new_segment_rubber_band.addPoint(self.cursor_point)
```

Comment:
This could be extracted to own method `self._undo_last_vertex()` to make it more obvious what the code does.